diff --git a/quickjs.c b/quickjs.c
index 48aeffc..5713a29 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -28204,7 +28204,9 @@ JSModuleDef *JS_RunModule(JSContext *ctx, const char *basename,
     return m;
 }
 
-static JSValue js_dynamic_import_job(JSContext *ctx,
+JSValue async_js_dynamic_import_job(JSContext *ctx,
+                                     int argc, JSValueConst *argv);
+JSValue js_dynamic_import_job(JSContext *ctx,
                                      int argc, JSValueConst *argv)
 {
     JSValueConst *resolving_funcs = argv;
@@ -28279,7 +28281,7 @@ static JSValue js_dynamic_import(JSContext *ctx, JSValueConst specifier)
     args[2] = basename_val;
     args[3] = specifier;
     
-    JS_EnqueueJob(ctx, js_dynamic_import_job, 4, args);
+    JS_EnqueueJob(ctx, async_js_dynamic_import_job, 4, args);
 
     JS_FreeValue(ctx, basename_val);
     JS_FreeValue(ctx, resolving_funcs[0]);
